### PR TITLE
AK: Make Function a little bit constexpr

### PR DESCRIPTION
**Pull request by nico**
Just enough to make it possible to pass a lambda to a constexpr function
taking an AK::Function parameter and have that constexpr function call
the passed-in AK::Function.

The main thing is that bit_cast<>s of pointers aren't ok in constexpr
functions, and neither is placement new. So add a union with stronger
typing for the constexpr case.

The body of the `if (is_constexpr_evaluated())` in
`init_with_callable()` is identical to the `if constexpr` right
after it. But `if constexpr (is_constexpr_evaluated())` always
evaluates `is_constexpr_evaluated()` in a constexpr context, so
this can't just add ` || is_constexpr_evaluated()` to that
`if constexpr`.

### CODE REVIEW

File: AK/Atomic.h [hunk 1]
```diff
@@ -262,16 +272,36 @@ class Atomic<T, DefaultMemoryOrder> {
         return __atomic_fetch_add(&m_value, val, order);
     }
 
+    ALWAYS_INLINE constexpr T fetch_add(T val, MemoryOrder order = DefaultMemoryOrder) noexcept
+    {
+        if (is_constant_evaluated()) {
+            int old_value = m_value;
```
Comment (spholz) [inline AK/Atomic.h:278]: Shouldn't this be `T` instead of `int`? (same for `fetch_sub`)

**Comment by nico** [comment]
I fixed the `int`.

I also made it so that the non-volatile overloads of `fetch_add`, `fetch_sub`, `load` call the volatile overloads. I don't know if that's necessary. Before we'd also call the volatile overload since there was no non-volatile one. After, we'd usually don't, before this tweak. I don't know if `this` being volatile is important for anything, but this way it's pretty certain that it doesn't change behavior.

@ADKaster @alimpfard Maybe you know more about this ^ :) (Before the current version of the patch, instead of the static_cast to volatile, I had just copied the bodies of the volatile versions in.)

**Comment by ADKaster** [comment]
:thonk: I don't know what the effect of adding non-volatile overloads of operator++ and such is. I also don't know what the effect of having volatile overloads is... volatile-casting this as you've done is probably going to get the same behavior in the end though.

**Comment by nico** [comment]
There's https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2024/p3309r0.html for std::atomic, and libc++ has both volatile and non-volatile versions that both just forward to a function without adjusting the load from `this`: https://github.com/llvm/llvm-project/blob/main/libcxx/include/__atomic/atomic.h#L167 (not constexpr yet though). So I'm guessing that in practice the previous version is probably fine too, but the new version should definitely be behavior preserving. So let's start with that, I suppose.

**Comment by nico** [comment]
Oh I should CTAD this, thanks. Done.

**Comment by nico** [comment]
> https://github.com/llvm/llvm-project/blob/main/libcxx/include/__atomic/atomic.h#L51

Looks like this one was mostly a lines-of-code savings thing when it was added: https://github.com/llvm/llvm-project/commit/776cc6e48f6954988bfb86c1854c14ba93d86c5e It's no longer saving lines of code now, so it doesn't have a great reason to be like that anymore.

**Comment by nico** [comment]
> And in some cases, they guard the volatile overload on is_always_lock_free (in this case, `atomic<T*>`)
> 
> https://github.com/llvm/llvm-project/blob/main/libcxx/include/__atomic/atomic.h#L443

These are from #67799 (support for floating point atomics), but I'm not sure why.

**Pull request merged**